### ISSUE TITLE
Fitting 3D mesh to image

### ISSUE THREAD

**Issue by zhangy76** [facebookresearch/pytorch3d#774]
I wrote a simple program for fitting a 3D model to image. The program minimizes the differences between rendered image and a target image over 3D mesh model parameters, texture parameters, and lighting parameters. The problem is only lighting parameters have gradient during back propagation. Below is the code
```
# define the parameters to be optimized
beta = torch.zeros([1,10]).to(device)
pose = nn.Parameter(torch.ones([1,48], requires_grad=True, device=device))
cams = nn.Parameter(torch.tensor([[0.,0.,0.5]], requires_grad=True, device=device))
lights_coeff = nn.Parameter(torch.zeros([1,9], requires_grad=True, device=device))
texture_coeff = nn.Parameter(torch.zeros([1,101], requires_grad=True, device=device))

# The optimizer
optimizer = torch.optim.SGD([pose,cams], lr=.1, momentum=0.9)

# define renderer
cameras = PerspectiveCameras(device=device, T=cams)

raster_settings = RasterizationSettings(
    image_size=224, 
    blur_radius=0.0, 
    faces_per_pixel=1, 
)

lights = DirectionalLights(device=device, ambient_color=lights_coeff[:,:3], diffuse_color=lights_coeff[:,3:6], direction=lights_coeff[:,6:])

renderer = MeshRenderer(
    rasterizer=MeshRasterizer(
        cameras=cameras, 
        raster_settings=raster_settings
    ),
    shader=SoftPhongShader(
        device=device, 
        cameras=cameras,
        lights=lights
    )
)

# update the parameters 
Niter = 100
for i in range(Niter):

    # obtain 3D vertices position given model parameters
    rotmat = batch_rodrigues(pose.view(-1,3)).view(-1,16,3,3)
    pred_output = smpl.forward(betas=beta, thetas_rotmat=rotmat)
    pred_vertices = pred_output.vertices
        
    # Add per vertex colors to texture the mesh
    new_tex_img = get_mano_texture(tex_basis, tex_mean, vec2texImg_index, texture_coeff)
    texture = TexturesUV(maps=new_tex_img, faces_uvs=faces_uvs.unsqueeze(0), verts_uvs=verts_uvs.unsqueeze(0))
    mesh = Meshes(pred_vertices, faces_idx, texture)

    lights = DirectionalLights(device=device, ambient_color=lights_coeff[:,:3], diffuse_color=lights_coeff[:,3:6], direction=lights_coeff[:,6:])
    
    predicted_rgb = renderer(mesh.clone(), T=cams)
    loss_rgb = ((predicted_rgb[0, ..., :3] - image_tensor) ** 2).mean()
    
    print('current loss: %f' % loss_rgb)
    # Optimization step
    optimizer.zero_grad()
    loss_rgb.backward(retain_graph=True)
    optimizer.step()
```
I have checked that the gradient of mesh model parameters can be propagated given a loss like "loss = ((pred_vertices - gt_vertices) ** 2).mean()". However, there is no gradient after adding the rendering parts

**Comment by bottler** [comment]
You need to move construction of `cameras` and `renderer` to inside the optimization loop. That should fix part of it. Can you  give us the definition of `get_mano_texture`?

**Comment by zhangy76** [comment]
Thank you very much for the reply. Do you mean I need to define the cameras and renderer every time of the updating? "texture_coeff" is the PCA coefficients of a texture model for the 3D mesh. "get_mano_texture" is a function to obtain texture map given the coeff.

**Comment by bottler** [comment]
> Do you mean I need to define the cameras and renderer every time of the updating?

Yes. A PyTorch3D object does not expect its member tensors to be modified underneath it. You generally need a new object if you want an object from new inputs. This is often designed to be fast.

--

Is get_mano_texture differentiable with respect to texture_coeff? Pasting its code might help us understand.

**Comment by zhangy76** [comment]
I moved the construction of cameras and renderer to inside the optimization loop. But it does not make any changes. 

def vec2img(vec, vec2texImg_index):
    # super slow and inefficient. Reorganize the code vector
    batch_size = vec.shape[0]
    img1d = torch.zeros([batch_size, 1024 * 1024 * 3]).cuda()
    img1d[:, vec2texImg_index] = vec
    return img1d.reshape((batch_size, 3, 1024, 1024)).permute((0, 3, 2, 1))

def get_mano_texture(tex_basis, tex_mean, vec2texImg_index, gamma):
    if not isinstance(gamma, torch.Tensor):
        gamma = torch.tensor(gamma, dtype=torch.float).cuda()
        
    scaled_gamma = gamma.permute(1, 0)
    offsets = torch.mm(tex_basis, scaled_gamma)
    tex_code = offsets + tex_mean[..., None]
    new_tex_img = vec2img(tex_code.permute(1, 0), vec2texImg_index) / 255

    return new_tex_img

I believe the process should be differentiable.

**Comment by yaadhavraajagility** [comment]
I am facing the same issue, were you able to resolve this?

**Comment by bottler** [comment]
@yaadhavraajagility please open a new issue with the details of your code and what you are experiencing. There isn't a bug here, something specific will be happening for you.

@zhangy76 In particular, what exactly are you seeing now? Which parameters are updating and which aren't.

**Comment by zhangy76** [comment]
For these four parameters pose ,cams,lights_coeff,texture_coeff, when I update them using torch.optim.SGD("the parameters", lr=.1, momentum=0.9), only lights_coeff have gradients.

**Comment by bottler** [comment]
I still don't know what you mean exactly by "have gradients". Could it be that everything is working fine but the renderer output actually has zero gradient with respect to the other parameters - e.g. because the camera is looking directly at a very boring part of the mesh? Can you visualize the renderer output with the original inputs, with the texture_coeff changed very slightly, and with the cams changed very slightly and check they all differ? (pose goes through more of your functions which we can't see.)

If they do, you could then work through each of the calculation steps individually and compare manual gradients to PyTorch3D gradients. I can't see how I could help any more from the information you have given.

**Comment by bottler** [comment]
I'm closing this for now because I don't think we can help without much more information.

**Comment by zhangy76** [comment]
Hi, sorry for late response.

Below are the rendered image and the reference image.
<img width="369" alt="rend" src="https://user-images.githubusercontent.com/60193243/128776475-d8617ae8-851b-48e6-9362-773e24c2d55b.png">
<img width="374" alt="ref" src="https://user-images.githubusercontent.com/60193243/128776476-28599dbb-9fa3-4126-955f-1550f8b4a7fe.png">

The gradients of the parameters I try to optimize are zeror; The loss of rendering differences does not change.

**Comment by zhangy76** [comment]
Found the problem. The function do not have gradients when the two images are too different.

**Comment by bfialkoff** [comment]
Did you manage to solve your issue? I'm working on something very very similar and I also cannot get it to converge, no matter what I do.
